pmc_heu.cpp

```diff
@@ -17,96 +17,116 @@
  ============================================================================
  */
 
+#include "pmc/pmc_bool_vector.h"
 #include "pmc/pmc_debug_utils.h"
 #include "pmc/pmc_heu.h"
 
 #include <algorithm>
 
 using namespace pmc;
-using namespace std;
 
 
-void pmc_heu::branch(vector<Vertex>& P, int sz,
-        int& mc, vector<int>& C, vector<short>& ind) {
+void pmc_heu::branch(std::vector<Vertex>& P, int sz,
+        int& mc, std::vector<int>& C, bool_vector& ind) {
 
-    if (P.size() > 0) {
+    if (!P.empty()) {
 
-        int u = P.back().get_id();
+        const int u = P.back().get_id();
         P.pop_back();
 
-        for (long long j = (*V)[u]; j < (*V)[u + 1]; j++)  ind[(*E)[j]] = 1;
+        for (long long j = (*V)[u]; j < (*V)[u + 1]; j++)  ind[(*E)[j]] = true;
 
-        vector <Vertex> R;
+        std::vector<Vertex> R;
         R.reserve(P.size());
-        for (int i = 0; i < P.size(); i++)
-            if (ind[P[i].get_id()])
-                if ((*K)[P[i].get_id()] > mc)
-                    R.push_back(P[i]);
 
-        for (long long j = (*V)[u]; j < (*V)[u + 1]; j++)  ind[(*E)[j]] = 0;
+        std::copy_if(P.begin(), P.end(), std::back_inserter(R),
+                     [this, mc, &ind](const Vertex &v) -> bool {
+                       return ind[v.get_id()] && (*K)[v.get_id()] > mc;
+                     });
 
-        int mc_prev = mc;
+        for (long long j = (*V)[u]; j < (*V)[u + 1]; j++)  ind[(*E)[j]] = false;
+
+        const int mc_prev = mc;
         branch(R, sz + 1, mc, C, ind);
 
         if (mc > mc_prev)  C.push_back(u);
 
-        R.clear();  P.clear();
+        P.clear();
     }
     else if (sz > mc)
         mc = sz;
     return;
 }
 
-int pmc_heu::search_bounds(pmc_graph& G,
-        vector<int>& C_max) {
+int pmc_heu::search_bounds(const pmc_graph& G,
```

Comment:
I made `G` `const` because it was really difficult to know what is mutated and what is not.